src/lib/zkapp.ts

```diff
@@ -826,6 +833,58 @@ Use the optional \`maxTransactionsWithActions\` argument to increase this number
       contract.account.sequenceState.assertEquals(actionsHash);
       return { state, actionsHash };
     },
+    getActions({
+      fromActionHash,
+      endActionHash,
+    }: {
+      fromActionHash?: Field;
+      endActionHash?: Field;
+    }): A[][] {
+      let actionsForAccount: A[][] = [];
+
+      Circuit.asProver(() => {
+        // if the fromActionHash is the empty state, we fetch all events
+        fromActionHash = fromActionHash
+          ?.equals(Events.emptySequenceState())
+          .toBoolean()
+          ? undefined
+          : fromActionHash;
+
+        // used to determine start and end values in string
+        let start: string | undefined = fromActionHash
+          ? Ledger.fieldToBase58(fromActionHash)
+          : undefined;
+        let end: string | undefined = endActionHash
+          ? Ledger.fieldToBase58(endActionHash)
+          : undefined;
+
+        let actions;
+
+        actions = Mina.getActions(contract.address, contract.self.tokenId);
+
+        // gets the start/end indices of our array slice
+        let startIndex = start
+          ? actions.findIndex((e) => e.hash === start) + 2
```

Comment:
the `+2` was a weird workaround.. I did that because the on-chain `actionHash` was always an action "behind", so I had to skip one

Reply:
let's fix computing action hashes (if actions are empty) and see if the workaround is still needed then
i would expect `+1`, because the hash stored on `actions` is the hash _after_ applying those actions, but with fromActionsHash we mean the hash _before_ the whole list of actions